[SPARK-31418][SCHEDULER] Request more executors in case of dynamic allocation is enabled and a task becomes unschedulable due to spark's blacklisting feature.

venkata91 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
In this change, when dynamic allocation is enabled instead of aborting immediately when there is an unschedulable taskset due to blacklisting, pass an event saying `SparkListenerUnschedulableTaskSetAdded` which will be handled by `ExecutorAllocationManager` and request more executors needed to schedule the unschedulable blacklisted tasks. Once the event is sent, we start the abortTimer similar to [SPARK-22148][SPARK-15815] to abort in the case when no new executors launched either due to max executors reached or cluster manager is out of capacity.
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
This is an improvement. In the case when dynamic allocation is enabled, this would request more executors to schedule the unschedulable tasks instead of aborting the stage without even retrying upto spark.task.maxFailures times (in some cases not retrying at all). This is a potential issue with respect to Spark's Fault tolerance.
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce any user-facing change?
No
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->


### How was this patch tested?
Added unit tests both in ExecutorAllocationManagerSuite and TaskSchedulerImplSuite
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

venkata91 committed:
[SPARK-31418][SCHEDULER] Request more executors in case of dynamic allocation is enabled and a task becomes unschedulable due to spark's blacklisting feature.

In this change, in the case of dynamic allocation is enabled instead of aborting an unschedulable blacklisted task blacklist immediately using the SparkListener pass an event saying UnschedulableBlacklistTaskSubmitted which will be handled by ExecutorAllocationManager and request more executors to schedule the unschedulable blacklisted task. Once the event is sent, we start the abortTimer similar to [SPARK-22148][SPARK-15815]

Currently manually tested it in our clusters. Also trying to figure out how to add unit tests.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

venkata91 commented:
Can you please review @squito @tgravescs @mridulm ?

tgravescs commented:
so took a quick look and I don't follow.  You send a message that says all blacklisted, but all you do is call the onSchedulerBacklogged, how does that add another executor if they are all blacklisted?  that will update when it needs to add more, but it doesn't change the number if calculated.  So for instance say I needed 3 executors. All of them blacklisted, updating the time to get a new one won't change that it thinks it  needs 3. 
If we change that to tell it to get one more, then my question is how do you do the proper accounting on that. This is why in the jira and other PR we said the allocation manager needs to be closely tied with the blacklist manager.

venkata91 commented:
> so took a quick look and I don't follow. You send a message that says all blacklisted, but all you do is call the onSchedulerBacklogged, how does that add another executor if they are all blacklisted? that will update when it needs to add more, but it doesn't change the number if calculated. So for instance say I needed 3 executors. All of them blacklisted, updating the time to get a new one won't change that it thinks it needs 3.
> If we change that to tell it to get one more, then my question is how do you do the proper accounting on that. This is why in the jira and other PR we said the allocation manager needs to be closely tied with the blacklist manager.

Ok. Got it, makes sense now. Let me think more about it.

venkata91 commented:
@tgravescs https://github.com/apache/spark/pull/22288 I went through the discussions as part of this PR. Is there any other PR you're referring to with discussions around Dynamic allocation and BlacklistManager? Also can you please explain a bit more on the proper accounting? Lets say we're requesting one executor (its possible, we're doing multiple rounds of requesting one executor) and also ensuring its with in the bound, won't dynamic allocation eventually get the num executors to consistent state with idle executors getting removed periodically?

tgravescs commented:
it might have been discussions no on PR or buried in comments, I don't have time to go looking.  there are many different conditions to consider. The main one we were focusing on is that you had as many executors as you needed to execute the task you had left.  This means the allocation manager was not going to ask for more. The problem is that some or all of those executor can get blacklisted.  The only way for the dynamic allocation manager to know it needs to ask for more is for it to know that nodes are blacklisted and it needs to ask for some more executors - thus internally incrementing is count of executors needed and asking yarn or other resource manager for more.  so now the number of executors the allocation manager is different then what its normal calculations would figure out.  simplified: #executors = (#tasks * #cpus per task/#cores per executor).   So you have to change the number of executors and you have to keep taking that it account because the allocation manager is always trying to calculate if it needs more or less executors.  You also have to notify it when executors become unblacklisted, or perhaps the ones that were blacklisted idle timeout, etc.  The allocation manager has to know a lot more details about the blacklisting and take that into account when its calculating the number of executors it needs.

venkata91 committed:
better log message

venkata91 committed:
Added one more test

venkata91 commented:
@tgravescs After thinking about the problem and also after discussing with @mridulm, I have handled this problem now by just keeping track of unschedulable task sets in order to add more executors when dynamic allocation is enabled. Now once some task becomes schedulable, we'll clear this set since some executor got free or we have just acquired a new executor and found a way to make progress. Let me know what do you think about this change. Thanks for taking a look previously and giving the overall context

venkata91 committed:
fix to avoid the corner case where we might not request any executor

venkata91 commented:
Can someone help me with why this Generate documents check is failing? Not sure I'm understanding the issue here. Any pointers would be appreciated.

tgravescs commented:
sorry haven't had a chance to look at your rework, I rekicked the checks as it might have been transient issue.

venkata91 committed:
Handle corner cases

venkata91 commented:
No worries @tgravescs Thanks for taking a look. For some reason, these checks keeps failing but it doesn't look to be related to my changes. some cache issue probably?

tgravescs commented:
yes its possible, I saw some issues on a few other pro although it was with different parts. It maybe early next week before I can review.

venkata91 commented:
> yes its possible, I saw some issues on a few other pro although it was with different parts. It maybe early next week before I can review.

Thanks that should be fine.

xkrogen commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -289,13 +290,23 @@ private[spark] class ExecutorAllocationManager(
+        tasksPerExecutor).toInt
+      math.max(totalNeed, executorMonitor.executorCountWithResourceProfile(rpId)) +
+        maxNeededForUnschedulables
+    } else {
```
I think this should be:
```
math.max(totalNeed, executorMonitor.executorCountWithResourceProfile(rpId) + maxNeededForUnschedules)
```
If we were already going to request `maxNeededForUnschedulables` _more_ executors than we currently have as part of `totalNeed`, that already satisfies the requirements.

mridulm replied:
Can you update with details about how this was resolved ? The original looked fine to me, but I want to make sure I am not missing something.

mridulm replied:
Scratch that - what I had seen was probably a modified version, not original ...

xkrogen commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -289,13 +290,23 @@ private[spark] class ExecutorAllocationManager(
     val maxNeeded = math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
       tasksPerExecutor).toInt
-    if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
+    val totalNeed = if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
```
I think `totalNeed` is a little confusing. Maybe `maxNeededWithSpeculationLocalityOffset` or something else more descriptive like that?

xkrogen commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -289,13 +290,23 @@ private[spark] class ExecutorAllocationManager(
+    if (numUnschedulables > 0) {
+      val maxNeededForUnschedulables = math.ceil(numUnschedulables * executorAllocationRatio /
+        tasksPerExecutor).toInt
+      math.max(totalNeed, executorMonitor.executorCountWithResourceProfile(rpId)) +
```
We discussed offline that using the current executor count can result in too many executors getting allocated, if the scheduling loop runs between when an executor is allocated and when tasks are scheduled on it (thus clearing out `unschedulableTaskSets`). Did you come up with any solution for this? If not and we think it's a small enough issue to ignore, let's at least add a comment about this caveat.

venkata91 replied:
Currently ExecutorAllocationManager runs every 100ms and resourceOffer happens every 1sec. `unschedulableTaskSets` gets cleared only if any task becomes schedulable and also we'll send the `unschedulable` event only once unless its removed from the `unschedulableTaskSets`. So I'm guessing it should be fine. Please correct me if I'm forgetting something.

venkata91 committed:
addressed Erik's comments

mridulm commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -789,6 +811,23 @@ private[spark] class ExecutorAllocationManager(
+            unschedulableTaskSets.add(stageAttempt)
+          case (None, _) =>
+            // Clear unschedulableTaskSets since atleast one task becomes schedulable now
+            unschedulableTaskSets.clear()
```
Why are we clearing all tasksets ? Shouldn't this not remove the specific taskset which became schedulable ?

venkata91 replied:
https://github.com/apache/spark/blob/263f04db865920d9c10251517b00a1b477b58ff1/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L670 - Based on the previous implementation, we want to defer killing taskSets as long as we have a non blacklisted executor which can be used to schedule a task. The problem is if we can only schedule only one of the task set then the other taskset might expire and the stage gets aborted. This is the reason I think why they cleared the expiry in the previous implementation which is why we are also clearing the `unschedulableTaskSets`

mridulm replied:
There is a difference in both cases.
If all tasksets are removed from `TaskSchedulerImpl.unschedulableTaskSetToExpiryTime`, it would defer aborting unschedulable tasksets in the assumption that some progress is being made and eventually tasksets might become scheduleable (or gets readded subsequently).
On other hand, if an unschedulable taskset remains in `unschedulableTaskSets`, this can trigger a successful executor allocation sooner - which could make it (and other taskset's) schedulable.

If we clear `unschedulableTaskSets` here, an eventual resourceOffer will result in re-addition to `unschedulableTaskSets` after a delay. If we want to defer resource acquisition, that sounds fine - even with this, current PR is still a strict improvement over what we have.

tgravescs replied:
Yeah I could go either way on this as well. if we only remove the task set it could help get more faster if lots of blacklisted or you only requested a few to begin with. I think its a big out of the norm for us to rely on the parameters here being None as well, traditionally we would have 2 messages, 1 for it being added, one for it being removed.  perhaps onUnschedulableTaskSetAdded and onUnschedulableTaskSetRemoved.

mridulm commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -829,12 +868,25 @@ private[spark] class ExecutorAllocationManager(
       numTotalTasks - numRunning
     }
 
+    def pendingUnschedulableTasksPerResourceProfile(rp: Int): Int = {
```
The method is returning number of task sets which are not schedulable - not number of tasks which are not scheduable.
`maxNumExecutorsNeededPerResourceProfile` is relying on this being task count - we will need to fix it here and in the event being fired.

venkata91 replied:
In the blacklisting logic, https://github.com/apache/spark/blob/263f04db865920d9c10251517b00a1b477b58ff1/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L615 - here we just check for the first task which is currently unschedulable (note: there could be more which we would eventually come back to), which is why we will only have 1 unschedulable task per taskset, thats why only we keep track of the unschedulableTaskSets and not the individual tasks in the taskSets themselves. So in short # of unschedulableTaskSets = # of unschedulableTasks fr now.

venkata91 replied:
I will refactor the name of this method to make it clear.

mridulm replied:
We should either base additional allocations on:
* Number of unschedulable tasks - this could potentially result in overallocation
* Do something similar to speculative tasks - so that there is atleast one executor allocation to make progress.

Currently, we are using unschedulable tasksets as a proxy for number of unschedulable tasks - which is a hybrid of both of these approaches.

Also note that given the discussion on clearing `unschedulableTaskSets`, we could toggle between asking for more resources and asking for less (as soon as a taskset becomes schedulable).

venkata91 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -696,7 +713,9 @@ private[spark] class ExecutorAllocationManager(
-        if (stageAttemptToNumTasks.isEmpty && stageAttemptToNumSpeculativeTasks.isEmpty) {
+        if (stageAttemptToNumTasks.isEmpty &&
+          stageAttemptToNumSpeculativeTasks.isEmpty &&
+          unschedulableTaskSets.isEmpty) {
```
Yes, thats right. Makes sense, will fix it.

venkata91 committed:
Addressed Mridul's comments

venkata91 commented on core/src/main/java/org/apache/spark/SparkFirehoseListener.java:
```diff
@@ -162,6 +162,11 @@ public void onSpeculativeTaskSubmitted(SparkListenerSpeculativeTaskSubmitted spe
     onEvent(speculativeTask);
   }
 
+  public void onUnschedulableBlacklistTaskSubmitted(
```
Done

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -311,6 +311,16 @@ private[spark] class DAGScheduler(
+   * Called by the TaskSetManager when there is an unschedulable blacklist task and dynamic
+   * allocation is enabled
+   */
+  def unschedulableBlacklistTaskSubmitted(
```
I think we should rename.   Maybe just remove Submitted.  since this isn't really Task Submission

venkata91 replied:
I thought I removed everything, seems I missed this one. Thanks for pointing out.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -829,12 +867,28 @@ private[spark] class ExecutorAllocationManager(
 
+    // Currently TaskSetManager.getCompletelyBlacklistedTaskIfAny only takes the first
+    // unschedulable task due to blacklisting. So keeping track of unschedulableTaskSets
+    // should be enough as we'll always have no more than a task unschedulable at any time.
```
I think this comment is confusing, specifically "as we'll always have no more than a task unschedulable at any time".
That isn't true. you could have lots of tasks blacklisted and thus unschedulable at any time. That function only returns the first one it finds.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -289,13 +290,26 @@ private[spark] class ExecutorAllocationManager(
+
+    // Since the maxNeededWithSpeculationLocalityOffset already includes the num executors needed
+    // to run unschedulable tasks, we would only try to add when the
+    // maxNeededWithSpeculationLocalityOffset is lesser than the active executors
```
this comment is also confusing to me.  its also above a check for numUnschedulables, which has nothing to with the comment.  Can we move it down next to the check on line 308 and clarify. I think you mean we don't add any tasks for the unschedulables if active executors is less than the number we are already asking for.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -789,6 +810,23 @@ private[spark] class ExecutorAllocationManager(
     }
 
+    override def onUnschedulableTaskSet
+      (unschedulableTaskSet: SparkListenerUnschedulableTaskSet): Unit = {
```
nit, fix indentation and style. "(" should be on previous line, parameters indented 4 spaces

tgravescs reviewed (commented):
At a high level seems like an ok approach to request at least some executors, even if it won't fit all of them, it makes it so you can make progress. Its unfortunate to add yet more tracking of the same thing in multiple places. I wish the allocation manager would move into scheduler, but that is much bigger change.

venkata91 committed:
Addressed tgraves and mridulm comments

tgravescs commented:
ok to test

SparkQA commented:
**[Test build #125613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125613/testReport)** for PR 28287 at commit [`0784dc3`](https://github.com/apache/spark/commit/0784dc3d8f5419822b2e26f2c1144e0a2e285a9e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30226/
Test PASSed.

SparkQA commented:
**[Test build #125613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125613/testReport)** for PR 28287 at commit [`0784dc3`](https://github.com/apache/spark/commit/0784dc3d8f5419822b2e26f2c1144e0a2e285a9e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
  * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
  * `case class SparkListenerUnschedulableTaskSetAdded(`
  * `case class SparkListenerUnschedulableTaskSetRemoved(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125613/
Test FAILed.

venkata91 commented:
> **[Test build #125613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125613/testReport)** for PR 28287 at commit [`0784dc3`](https://github.com/apache/spark/commit/0784dc3d8f5419822b2e26f2c1144e0a2e285a9e).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds the following public classes _(experimental)_:
> * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
> * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
> * `case class SparkListenerUnschedulableTaskSetAdded(`
> * `case class SparkListenerUnschedulableTaskSetRemoved(`

It seems like the failed tests are unrelated seems like transient issues. @tgravescs Should we rerun it again?

tgravescs commented:
test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30250/
Test PASSed.

SparkQA commented:
**[Test build #125637 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125637/testReport)** for PR 28287 at commit [`0784dc3`](https://github.com/apache/spark/commit/0784dc3d8f5419822b2e26f2c1144e0a2e285a9e).

SparkQA commented:
**[Test build #125637 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125637/testReport)** for PR 28287 at commit [`0784dc3`](https://github.com/apache/spark/commit/0784dc3d8f5419822b2e26f2c1144e0a2e285a9e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
  * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
  * `case class SparkListenerUnschedulableTaskSetAdded(`
  * `case class SparkListenerUnschedulableTaskSetRemoved(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125637/
Test FAILed.

SparkQA commented:
**[Test build #125723 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125723/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30336/
Test PASSed.

SparkQA commented:
**[Test build #125723 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125723/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
  * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
  * `case class SparkListenerUnschedulableTaskSetAdded(`
  * `case class SparkListenerUnschedulableTaskSetRemoved(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125723/
Test FAILed.

venkata91 commented:
> **[Test build #125723 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125723/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds the following public classes _(experimental)_:
> * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
> * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
> * `case class SparkListenerUnschedulableTaskSetAdded(`
> * `case class SparkListenerUnschedulableTaskSetRemoved(`

@tgravescs it seems like it again failed on a transient issue on one of the test. Shall we rerun this one?

tgravescs commented:
test this please

SparkQA commented:
**[Test build #125783 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125783/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30394/
Test PASSed.

SparkQA commented:
**[Test build #125783 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125783/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
  * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
  * `case class SparkListenerUnschedulableTaskSetAdded(`
  * `case class SparkListenerUnschedulableTaskSetRemoved(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125783/
Test FAILed.

venkata91 commented:
> **[Test build #125783 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125783/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds the following public classes _(experimental)_:
> * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
> * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
> * `case class SparkListenerUnschedulableTaskSetAdded(`
> * `case class SparkListenerUnschedulableTaskSetRemoved(`

It seems like it fails every time with different test and everyone of them seems transient. @tgravescs I saw a recent change by @HyukjinKwon `SPARK-32245` to run Spark tests as Github actions primarily addressing the transient issues due to jenkins machines slowness. Do you think we should try running these tests using Github Actions? I am not sure how does that work. It would be great if you can help here.

tgravescs commented:
I'm not sure if you can pick it up, but kicked the GitHub Actions again to see.

tgravescs commented:
test this please

tgravescs commented:
I still need to look at your changes as well, haven't had time yet

SparkQA commented:
**[Test build #125838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125838/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30449/
Test PASSed.

SparkQA commented:
**[Test build #125838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125838/testReport)** for PR 28287 at commit [`b4c27fb`](https://github.com/apache/spark/commit/b4c27fb2539d10356bb4208e0509f7a32a577613).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnschedulableTaskSetAdded(stageId: Int, stageAttemptId: Int)`
  * `case class UnschedulableTaskSetRemoved(stageId: Int, stageAttemptId: Int)`
  * `case class SparkListenerUnschedulableTaskSetAdded(`
  * `case class SparkListenerUnschedulableTaskSetRemoved(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125838/
Test PASSed.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -829,12 +873,28 @@ private[spark] class ExecutorAllocationManager(
+
     def hasPendingTasks: Boolean = {
-      hasPendingSpeculativeTasks || hasPendingRegularTasks
+      hasPendingSpeculativeTasks || hasPendingRegularTasks || hasPendingUnschedulableTasks
```
this doesn't make sense to me. Did you have a specific bug without this?  
The tasks themselves are counted as regular tasks.

venkata91 replied:
No, this is not needed. I will remove this. Thanks.

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -311,6 +311,26 @@ private[spark] class DAGScheduler(
 
+  /**
+   * Called by the TaskSetManager when a taskset becomes unschedulable due to blacklisting and
+   * dynamic allocation is enabled
```
nit, add period at end

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -311,6 +311,26 @@ private[spark] class DAGScheduler(
+
+  /**
+   * Called by the TaskSetManager when an unschedulable taskset becomes schedulable and dynamic
+   * allocation is enabled
```
nit add period

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1014,6 +1034,20 @@ private[spark] class DAGScheduler(
+      stageId: Int,
+      stageAttemptId: Int): Unit = {
+    listenerBus.post(
+      SparkListenerUnschedulableTaskSetAdded(stageId, stageAttemptId))
```
I think this will fit on previous line

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala:
```diff
@@ -339,6 +349,20 @@ private[spark] trait SparkListenerInterface {
 
+  /**
+   * Called when a taskset becomes unschedulable due to blacklisting and dynamic allocation
+   * is enabled
```
nit add period.

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala:
```diff
@@ -339,6 +349,20 @@ private[spark] trait SparkListenerInterface {
+
+  /**
+   * Called when an unschedulable taskset becomes schedulable and dynamic allocation
+   * is enabled
```
add period

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -696,7 +717,8 @@ private[spark] class ExecutorAllocationManager(
 
```
I think we should remove the stage attempt from unschedulableTaskSets when the stage completes in case it fails or something and never send the unschedulabletasksetremoved

venkata91 replied:
I think as long as the stageattempt is schedulable we can remove it, although it can fail but eventually if it becomes unschedulable again, we would add it back again to unschedulableTaskSets right? Wouldn't that be simpler?

tgravescs reviewed (commented):
it would also be nice to add a test to make sure the abort timer still works as expected. Please check TaskSchedulerImplSuite

venkata91 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -289,13 +290,27 @@ private[spark] class ExecutorAllocationManager(
     }
+
+    if (unschedulableTaskSets > 0) {
+      // Request additional executors only if maxNeededWithSpeculationLocalityOffset is less than
```
Thanks for the suggestion. This is much better.

venkata91 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -829,12 +873,28 @@ private[spark] class ExecutorAllocationManager(
 
+    // Currently TaskSetManager.getCompletelyBlacklistedTaskIfAny only takes the first unschedulable
+    // task found due to blacklisting. This way we only need to keep track of the unschedulable
+    // tasksets which is an indirect way to get the current number of unschedulable tasks.
```
Thanks for the suggestion!

venkata91 committed:
Address tgraves review comments

venkata91 commented:
> it would also be nice to add a test to make sure the abort timer still works as expected. Please check TaskSchedulerImplSuite

@tgravescs Do you mean a test where allocation manager is not able to acquire a new executor with in the timeout and the stage should get aborted?

tgravescs commented:
yes we have a test in TaskSchedulerImplSuite that checks to make sure it aborted, but I don't think it covers when dynamic allocation on, so it doesn't hit your new code.  So we would want to add a test where it can't acquire a new executor and aborts.

venkata91 commented:
> yes we have a test in TaskSchedulerImplSuite that checks to make sure it aborted, but I don't think it covers when dynamic allocation on, so it doesn't hit your new code. So we would want to add a test where it can't acquire a new executor and aborts.

I think the new test which I added is just duplicated. Do you think its better to just add the config to enable dynamic allocation to the other test itself in order to avoid the duplication

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30526/
Test PASSed.

tgravescs commented:
you can make a common function that has most of the code that gets called from 2 separate tests. one test passes with dynamic allocation on, the other with it off. that will reduce code duplication.

venkata91 committed:
fix the test changes

venkata91 commented:
> you can make a common function that has most of the code that gets called from 2 separate tests. one test passes with dynamic allocation on, the other with it off. that will reduce code duplication.

nevermind, I made some changes to the test so that it goes to the `None` block where we check if dynamic allocation is enabled or not and request accordingly.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30528/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30531/
Test PASSed.

SparkQA commented:
**[Test build #125918 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125918/testReport)** for PR 28287 at commit [`e167da5`](https://github.com/apache/spark/commit/e167da53b0b8ffde746c86e0564d7971363f3746).

SparkQA commented:
**[Test build #125919 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125919/testReport)** for PR 28287 at commit [`112bb3a`](https://github.com/apache/spark/commit/112bb3a3e72c664f2cbb5bebca0265f5e9c23303).

SparkQA commented:
**[Test build #125921 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125921/testReport)** for PR 28287 at commit [`0d07845`](https://github.com/apache/spark/commit/0d07845459e6e0f606e5f0920c973de256a309e1).

SparkQA commented:
**[Test build #125919 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125919/testReport)** for PR 28287 at commit [`112bb3a`](https://github.com/apache/spark/commit/112bb3a3e72c664f2cbb5bebca0265f5e9c23303).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125919/
Test FAILed.

SparkQA commented:
**[Test build #125921 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125921/testReport)** for PR 28287 at commit [`0d07845`](https://github.com/apache/spark/commit/0d07845459e6e0f606e5f0920c973de256a309e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125921/
Test PASSed.

SparkQA commented:
**[Test build #125918 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125918/testReport)** for PR 28287 at commit [`e167da5`](https://github.com/apache/spark/commit/e167da53b0b8ffde746c86e0564d7971363f3746).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125918/
Test FAILed.

tgravescs commented:
org.apache.spark.ExecutorAllocationManagerSuite failed in the GitHub Actions please take a look

tgravescs commented:
Sorry that failure might be related to https://issues.apache.org/jira/browse/SPARK-32287 I kicked it again

venkata91 commented:
@tgravescs it seems like it again failed somewhere in pyspark module. Not sure if this is transient or not.

tgravescs commented:
test this please

SparkQA commented:
**[Test build #125996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125996/testReport)** for PR 28287 at commit [`0d07845`](https://github.com/apache/spark/commit/0d07845459e6e0f606e5f0920c973de256a309e1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30610/
Test PASSed.

SparkQA commented:
**[Test build #125996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125996/testReport)** for PR 28287 at commit [`0d07845`](https://github.com/apache/spark/commit/0d07845459e6e0f606e5f0920c973de256a309e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125996/
Test PASSed.

venkata91 commented:
phew! Finally the tests are passing

venkata91 commented:
@tgravescs Can you please take a look again? Now that the tests are passing. Thanks

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -829,6 +872,23 @@ private[spark] class ExecutorAllocationManager(
+      attempts.filter(attempt => unschedulableTaskSets.contains(attempt)).size
+    }
+
+    def hasPendingUnschedulableTasks: Boolean = {
```
This isn't used anymore, correct?  Remove it if not

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1014,6 +1034,18 @@ private[spark] class DAGScheduler(
+  }
+
+  private[scheduler] def handleUnschedulableTaskSetRemoved(
+    stageId: Int,
```
indentation here should be 4 spaces

tgravescs reviewed (commented):
couple minor things otherwise looks good

venkata91 committed:
Address tgraves review comments

venkata91 commented:
> couple minor things otherwise looks good

Awesome. Thanks. Addressed the review comments. Hopefully the tests should pass fine the first time itself.

SparkQA commented:
**[Test build #126282 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126282/testReport)** for PR 28287 at commit [`2f019d5`](https://github.com/apache/spark/commit/2f019d5dc1844eba53cdebabb7adfed8c911830c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30898/
Test PASSed.

SparkQA commented:
**[Test build #126283 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126283/testReport)** for PR 28287 at commit [`d9f473d`](https://github.com/apache/spark/commit/d9f473d748b206426106a9c4e97e153aec2f508f).

venkata91 commented:
@tgravescs it seems like couple of tests are failing. Can you please kick this off again?

SparkQA commented:
**[Test build #126282 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126282/testReport)** for PR 28287 at commit [`2f019d5`](https://github.com/apache/spark/commit/2f019d5dc1844eba53cdebabb7adfed8c911830c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126282/
Test PASSed.

SparkQA commented:
**[Test build #126283 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126283/testReport)** for PR 28287 at commit [`d9f473d`](https://github.com/apache/spark/commit/d9f473d748b206426106a9c4e97e153aec2f508f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126283/
Test PASSed.

tgravescs commented:
@venkata91 I assume your updates keep doing rebase?  If possible its great if you can just do up merges because doing rebases  makes seeing the diffs between  requested changes much harder.

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -311,6 +311,26 @@ private[spark] class DAGScheduler(
+   * allocation is enabled.
+   */
+  def unschedulableTaskSetRemoved(
+    stageId: Int,
```
indentation needs to be 4 spaces

venkata91 committed:
Addressed indentation review comments

venkata91 commented:
> @venkata91 I assume your updates keep doing rebase? If possible its great if you can just do up merges because doing rebases makes seeing the diffs between requested changes much harder.

Sure, makes sense. Will keep that in mind for the future.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30968/
Test PASSed.

SparkQA commented:
**[Test build #126352 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126352/testReport)** for PR 28287 at commit [`d6f1e73`](https://github.com/apache/spark/commit/d6f1e73ae49d9641a8efcb1f5e9016bd378a70c6).

tgravescs reviewed (approved):
pending Jenkins, looks good.

SparkQA commented:
**[Test build #126352 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126352/testReport)** for PR 28287 at commit [`d6f1e73`](https://github.com/apache/spark/commit/d6f1e73ae49d9641a8efcb1f5e9016bd378a70c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126352/
Test PASSed.

tgravescs commented:
merged to master, thanks @venkata91

venkata91 commented:
> merged to master, thanks @venkata91

Thanks @tgravescs for patiently doing multiple rounds of reviews. :)
